### ISSUE TITLE
Improve JWT Performance

### ISSUE THREAD

**Issue by jdmcclur** [OpenLiberty/open-liberty#20171]
While running one of my JWT primitive performance throughput tests, I see the following bottleneck:

https://github.com/OpenLiberty/open-liberty/blob/4500c8a6c6fe8b37142176aacb895640ca751f87/dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java#L899

```
  Parent      0   0.00   9.56           0         680    J:com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.performDelegation(Ljava/lang/String;)V
 
    Self      0   0.00   **9.56**           0         680    J:com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.getSubjectToString(Ljavax/security/auth/Subject;)Ljava/lang/String;
 
   Child      0   0.00   9.56           0         680    J:java/security/AccessController.doPrivileged(Ljava/security/PrivilegedAction;)Ljava/lang/Object;
```

We are  doing a toString on the Subject, which has to go through a doPriv and is pretty slow (taking almost 10% of the time in this test).

My question is: do we really have to do a toString on the whole Subject? It looks like this code is looking for an "accessId" string and the user associated with it. Could we just look at the Credentials instead of the whole Subject?  

I coded up a "proof of concept" to look through the credentials that helped shave off 8 to 9% of the overhead (in my case), but not sure if we would need to sync things, or add a doPriv, or if we even need to look at both the public and private credentials. I don't understand everything going on here.
https://github.com/jdmcclur/open-liberty/commit/98271b928c447560da21e655a4491b536f53d65b

**Comment by jhanders34** [comment]
This issue actually highlights a bigger problem that I have seen in the past when reducing the amount of work done during requests due to audit behavior.  I had seen this in the past with another collaborator and made an update in order to avoid some of this extra work by doing it in the audit function.  If the Audit class had a isRequired(EventID) that queried the AuditService to see if audit is enabled for that event id, we could guard all of the code that does audit function.  If we cannot get that function, then i think the best approach is to pass an Object that contains the Subjects and then does the processing in the AuditPE function to convert it to the Strings that are being looked for.  This way if the audit feature is not enabled, we do not pay this cost.  I do think that a Audit.isRequired method though would be the best option.

**Comment by jhanders34** [comment]
I guess the other option is to set the AuditService in each place that uses Audit in order to call isRequired directly on the AuditService.

**Comment by jhanders34** [comment]
Looking back at some of my changes, it looks like the Web collaborator also has the same problem I fixed in the EJB collaborator that can be easily improved upon with a simple PR like the one I did:  #3517.  For reference I did #14596 to delay formatting until processing the audit that I was referring to above. These are all bandaids that are necessary for when audit is enabled to improve audit enabled performance, but to improve the default case for when it is not enabled still will have the most benefit from a call to isRequired to avoid all the processing.

**Comment by una-tapa** [comment]
Thanks Jared. Putting this high priority.

**Comment by jhanders34** [comment]
To give some more details about what was done to resolve this specific defect:  I updated to add a isAuditRequired method that is used to not process things that will do nothing anyway.  Additionally I did put in similar to changes to #3517 and other changes to correctly do things compared to what we were doing or doing things more performant.  Follow on PRs will be done later to increase use of isAuditRequired and reduce some of the expense of getting the delegated users without using Subject.toString.  I am doing the former and the security team is looking at the latter.
